### PR TITLE
feat(comm/attn_offload.py): support selective ckpt and cpu offload

### PR DESCRIPTION
**Pull request by huangting4201**
## Motivation

Support selective checkpoint and cpu offload asynchronously to improve performance.

1. When enabling checkpointing, if `selective_checkpoint` is set to True, for layers that are recomputed, storing the intermediate activations of the **attention** part allows the attention part to be skipped during recomputation, thereby enhancing training performance.
2. If `selective_checkpoint_offload` is further set to True, the intermediate activations of the attention part for the recomputed layers will be asynchronously offloaded to the CPU to save GPU memory.

However, it should be noted that current testing has revealed that when `selective_checkpoint_offload` is set to True, the `DtoH` and `HtoD` operations compete with `allgather` and other communications for bandwidth, leading to increased `allgather` communication times and a consequent decline in overall performance. Therefore, it is advisable to avoid enabling selective_checkpoint_offload when it is not necessary.

## Modification

`internlm/core/parallel/comm/attn_offload.py`: AttnOffloadManager, a manager for attention output CPU offloading and GPU prefetch loading.

## Use cases (Optional)

example config：
```
selective_checkpoint = True
selective_checkpoint_offload = False
model = dict(
    num_chunks=1,  # if num_chunks > 1, interleaved pipeline scheduler is used.
    checkpoint=1,  # The proportion of layers for activation aheckpointing, the optional value are True/False/[0-1]
    ......
)
```
note：**should be used with isp, and only GQA is supported now**

**loss accuracy checking**
img_v3_02gj_77259706-f94c-466c-a451-ee88aad1de7g

img_v3_02gj_49a5bb6d-af96-4af9-9bc3-49493ad86fdg

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: internlm/core/parallel/comm/attn_offload.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    """
+
+    def __init__(self, enable_cpu_offload: bool = False) -> None:
+        # cpu offload overlapping
+        self.cpu_offload = enable_cpu_offload
+        # layer id mapping to flash attn output
+        self.fa_output_mapping = {}
+        self.fa_stream = torch.cuda.Stream()
+        self.d2h_final_event = torch.cuda.Event()
+        self.h2d_final_event = torch.cuda.Event()
```
Comment (yingtongxiong) [inline internlm/core/parallel/comm/attn_offload.py:20]: 这两个event在下面似乎没有被用到（被注释了），这里是留着以后配合下面的record和wait用？

File: internlm/core/parallel/comm/isp.py [hunk 2]
```diff
@@ -481,13 +493,36 @@ def _pre_forward_hook_for_out_proj(self, module: nn.Module, *args):  # pylint: d
...
+        # register offload and prefetch hook for selective ckpt with wo linear
+        if self.sc_offload is True:
+            # move current layer's attn output from GPU to CPU asynchronizely
+            if (
+                self.is_forward is True
+                and gpc.config.selective_checkpoint
+                and block_index not in self.layers_fa_not_release
+                and block_index < self._ckpt_block_num
+            ):
+                get_offload_manager().offload_fa_output_with_layer(layer_idx=block_index)
```
Comment (yingtongxiong) [inline internlm/core/parallel/comm/isp.py:530]: _pre_forward_hook_for_out_proj是注册在self._overlap_states[cid].isp_outs这个的pre-forward hook上，我理解这里首先得有attention的output才能offload，然后结合这个pr https://github.com/InternLM/InternEvo/pull/381， 如果self._overlap_states[cid].isp_outs为“wqkv“或者”attn"时，atten是不是还没开始算，这边就开始offload了。不知道这一块逻辑是如何处理的？

File: internlm/core/parallel/comm/attn_offload.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        tensors = []
+        for tensor_id, tensor in enumerate(output):
+            if tensor is None:
+                tensors.append(None)
+                continue
+            tensor_buf = self.get_tensor_buf_for_offloaded_tensor(tensor, layer_idx, tensor_id)
+            tensor_buf.copy_(tensor)
+            tensors.append(tensor_buf)
+        self.fa_output_mapping[layer_idx] = tensors
+        torch.cuda.current_stream(get_current_device()).synchronize()
```
Comment (yingtongxiong) [inline internlm/core/parallel/comm/attn_offload.py]: 这里加当前计算stream的synchronize的用处是？如果没有的话，会引起什么后果吗？

File: internlm/model/ops/_flash_attn.py [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+            normalization factor).
+        S_dmask [optional, if return_attn_probs=True]: (batch_size, nheads, seqlen, seqlen).
+            The output of softmax (possibly with different scaling). It also encodes the dropout
+            pattern (negative means that location was dropped, nonnegative means it was kept).
+    """
+
+    assert gpu_flash_attn_impl is True and flash_attn.__version__ in [
+        "2.2.1",
+        "2.6.3",
+    ], "flash-attn should be installed and version must be v2.2.1 or v2.6.3"
```
Comment (yingtongxiong) [inline internlm/model/ops/_flash_attn.py:298]: 为什么是指定这两个版本？有什么特殊考虑嘛

File: internlm/core/parallel/comm/attn_offload.py (hunk 1 above)
Comment (huangting4201) [reply internlm/core/parallel/comm/attn_offload.py:20]: 就是有两种方式，1.用event来做同步 2.用fa_stream sync。目前代码里用了第二种，但是第一种我暂时还不想删掉，所以注释掉留着了

File: internlm/model/ops/_flash_attn.py (hunk 4 above)
Comment (huangting4201) [reply internlm/model/ops/_flash_attn.py:298]: 这个是我们之前定好的版本

File: internlm/core/parallel/comm/attn_offload.py (hunk 3 above)
Comment (huangting4201) [reply internlm/core/parallel/comm/attn_offload.py]: 因为我们用fa stream做异步offload，所以这里在特定操作时比如save或者load时要保证主stream的操作是已经完成的，所以要加sync，不然会出现nan grad

File: internlm/core/parallel/comm/isp.py (hunk 2 above)
Comment (huangting4201) [reply internlm/core/parallel/comm/isp.py:530]: 这块确实会有问题，我之前考虑要加个assert，忘记加了

File: internlm/core/parallel/comm/attn_offload.py (hunk 1 above)
Comment (yingtongxiong) [reply internlm/core/parallel/comm/attn_offload.py:20]: ok

File: internlm/core/parallel/comm/attn_offload.py (hunk 3 above)
Comment (yingtongxiong) [reply internlm/core/parallel/comm/attn_offload.py]: 也就是说，异步的save和load之前得保证这个insert操作是完成的，否则save和load的对象可能是不完整的，从而造成nan grad

File: internlm/model/ops/_flash_attn.py (hunk 4 above)
Comment (yingtongxiong) [reply internlm/model/ops/_flash_attn.py:298]: ok

File: internlm/core/parallel/comm/attn_offload.py (hunk 3 above)
Comment (huangting4201) [reply internlm/core/parallel/comm/attn_offload.py]: 对的，因为我们是两个stream分开处理，所以我们得保证我们在做insert前，主stream的计算是完成的，才能保证数据是完整的

File: internlm/core/parallel/comm/isp.py (hunk 2 above)
Comment (huangting4201) [reply internlm/core/parallel/comm/isp.py:530]: TODO: 等https://github.com/InternLM/InternEvo/pull/381合入后，需要加assert判断，保证是wo处启动ag

File: internlm/core/parallel/comm/attn_offload.py (hunk 3 above)
Comment (yingtongxiong) [reply internlm/core/parallel/comm/attn_offload.py]: 好的

**Comment by huangting4201** [comment]
依赖pr https://github.com/InternLM/InternEvo/pull/381

File: internlm/core/parallel/comm/isp.py (hunk 2 above)
Comment (yingtongxiong) [reply internlm/core/parallel/comm/isp.py:530]: ok

File: internlm/core/parallel/comm/attn_offload.py (hunk 1 above)
Comment (mwiacx) [reply internlm/core/parallel/comm/attn_offload.py:20]: fa_stream sync需要会阻塞cpu，这可能会有些性能问题，需要主意一下

File: internlm/core/parallel/comm/isp.py (hunk 2 above)
Comment (huangting4201) [reply internlm/core/parallel/comm/isp.py:530]: Done

**Review by yingtongxiong — APPROVED** [review]

**Review by sunpengsdu — APPROVED** [review]
